addons/predicates/tkr.go

```diff
@@ -18,3 +22,38 @@ func TKR(log logr.Logger) predicate.Funcs {
 		GenericFunc: func(e event.GenericEvent) bool { return true },
 	}
 }
+
+// ClusterHasLabel checks if the cluster has the given label
+func ClusterHasLabel(label string, logger logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return processIfClusterHasLabel(label, e.ObjectNew, logger.WithValues("predicate", "updateEvent"))
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return processIfClusterHasLabel(label, e.Object, logger.WithValues("predicate", "createEvent"))
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return processIfClusterHasLabel(label, e.Object, logger.WithValues("predicate", "deleteEvent"))
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfClusterHasLabel(label, e.Object, logger.WithValues("predicate", "genericEvent"))
+		},
+	}
+}
+
+func processIfClusterHasLabel(label string, obj client.Object, logger logr.Logger) bool {
+	kind := obj.GetObjectKind().GroupVersionKind().Kind
+	cluster := &clusterapiv1beta1.Cluster{}
+
+	if kind == "" || kind != cluster.Kind {
```

Comment:
`cluster.Kind` is always ` == ""` here (based on `cluster := &clusterapiv1beta1.Cluster{}` in the above line).

We don't need to check for `kind == ""`: we simply need to get Cluster kind properly. We can do that, e.g. like this:
```go
var clusterKind = reflect.TypeOf(clusterapiv1beta1.Cluster{}).Name()
```
Which is best to do as a standalone var (not every time when you need it).

BTW, do we have a unit test for this? This function looks to be perfectly unit-testable.

Reply:
@imikushin I had to add this check for ```kind == "" ``` as the the tests for addons controller are using dynamic client to create resources from unstructured map created by parsing the yaml. https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/testutil/test_helpers.go#L60
After debugging, I found out that the kind field of the resource does not get populated there (and doesn't get updated if I try to update it later). So these check for the emptiness of the kind field was to cover those. Tried a lot but didn't find any way to fix that seemingly a bug in test utils

Reply:
You are right. I'll add unit test for this function.

Reply:
@imikushin You were right! !!```reflect.TypeOf(clusterapiv1beta1.Cluster{}).Name()``` is the correct way to go. I wonder there are so many usages of ```cluster := &clusterapiv1beta1.Cluster{}``` followed by ```cluster.Kind``` in our code base, but I just checked and the output would be an empty string. Not sure how those are working!

Reply:
yup, the Kind is indeed empty. Could you log an issue for some of the other controllers that are doing this?

Reply:
Sure @vijaykatam. Filed [this issue](https://github.com/vmware-tanzu/tanzu-framework/issues/1711)